pkg/controllers/nodeoverlay/store.go

```diff
@@ -0,0 +1,287 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeoverlay
+
+import (
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1alpha1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+	nodepoolutils "sigs.k8s.io/karpenter/pkg/utils/nodepool"
+)
+
+type PriceUpdate struct {
+	OverlayUpdate           *string
+	lowestWeightOverlayName string
+	lowestWeight            *int32
+}
+
+type CapacityUpdate struct {
+	OverlayUpdate           corev1.ResourceList
+	lowestWeightOverlayName string
+	lowestWeight            *int32
+}
+
+type InstanceTypeUpdate struct {
+	Price    map[string]*PriceUpdate
+	Capacity *CapacityUpdate
```

Comment:
Any reason to make these nillable when the fields inside are already nillable? Seems like it would simplify logic if these always existed once an InstanceTypeUpdate is created. See this section [here](https://github.com/kubernetes-sigs/karpenter/pull/2306/files?diff=unified&w=0#diff-667b9a6e671f00f1bce9dcc006346f54d99a200d70e6e07e3fcf3a69462427eeR108-R120), we should be able to collapse the two branches.